### ISSUE TITLE
Whether `react/display-name` requires forwardRefs to have `displayName` changes if function contains inline condition that returns an object

### ISSUE THREAD

**Issue by AlanSl** [yannickcr/eslint-plugin-react#3192]
I have a library with `forwardRef` components that all have `displayName` properties, apparently mandated by the linter; however, a few somehow slipped through without `displayName` being added, undetected by the linter. 

The pattern as to whether the linter would or would not view missing `displayName` as a problem is very odd. It looks like the pattern is, a `forwardRef`'s return value requires a `displayName` UNLESS the callback function contains an inline condition (`?`, `&&`, `||`...) where one or more of the possible values is an object.

Here's a simplified example that  **does not** require a `displayName` (with a ternary that may return an object):

```jsx
const Example = forwardRef(({ children }, ref) => {
  const props = children ? { testID: 'yes' } : { testID: 'no' }
  return (
    <View {...props} ref={ref}>
      {children}
    </View>
  )
})
```

![image](https://user-images.githubusercontent.com/29628323/151543743-2c153b77-9ec5-4750-bb48-29fa71cf659e.png)

...and here's an example that **does** require display name. It's functionally identical, but the ternary returns strings rather than objects, and for some reason that changes the linter rule behaviour:

```jsx
const Example = forwardRef(({ children }, ref) => {
  const props = { testID: children ? 'yes' : 'no' }
  return (
    <View {...props} ref={ref}>
      {children}
    </View>
  )
})
```

---

Here's another example that **does not** require display name, this time with `&&` instead of `?`:

```jsx
const Example = forwardRef(({ children }, ref) => {
  const style = children && { backgroundColor: 'blue' }
  return (
    <View style={style} ref={ref}>
      {children}
    </View>
  )
})
```

![image](https://user-images.githubusercontent.com/29628323/151543997-765138e0-40d0-40fe-9bf1-e13f5cc85bd9.png)

...and here's an example that is the same except the `&&` returns a string, not an object, so it **does** require a display name: 

```jsx
const Example = forwardRef(({ children }, ref) => {
  const testID = children && 'blue'
  return (
    <View testID={testID} ref={ref}>
      {children}
    </View>
  )
})
```

![image](https://user-images.githubusercontent.com/29628323/151544233-76a5ea23-1f22-4275-9d3e-518c7a5d3db9.png)

---

#### Versions

Tested and confirmed on `eslint-plugin-react@7.28.0` (current latest), first seen on `eslint-plugin-react@7.23.2`. 

Project is a yarn 1 monorepo with `eslint@7.32.0` (latest `7.x` version; `yarn why` finds only one version each for `eslint` and `eslint-plugin-react`).

Screenshots in VSCode 1.63.2 running on MacOS 11.6.1.

**Comment by ljharb** [comment]
I agree that these are bugs; all of these examples should be warned on.

(another simpler fix than adding `.displayName =`, ofc, is to never use arrow functions here, and only use named functions, but either way the linter should be warning)
